File: RoombAPI.Server/Services/BatteryService.cs
namespace RoombaPi.Server.Services
{
  using System.Threading.Tasks;
  using Create2.Sensors.Battery;
  using global::Grpc.Core;
  using Google.Protobuf.WellKnownTypes;
  using RoombaPi.Grpc;
  using ChargingState = RoombaPi.Grpc.ChargingState;

  public class BatteryService : Battery.BatteryBase
  {
    private readonly IBatterySensor batterySensor;

    public BatteryService(IBatterySensor batterySensor)
    {
      this.batterySensor = batterySensor;
    }

    public override async Task<BatteryInfoResponse> GetBatteryInfo(Empty request, ServerCallContext context)
    {
      return new BatteryInfoResponse
      {
        Capacity = await this.batterySensor.GetCapacity(),
        Charge = await this.batterySensor.GetCharge(),
        Current = await this.batterySensor.GetCurrent(),
        Temperature = await this.batterySensor.GetTemperature(),
        Voltage = await this.batterySensor.GetVoltage()
      };
    }

    public override async Task<ChargingSourcesAvailableResponse> GetChargingSourcesAvailable(Empty request,
      ServerCallContext context)
    {
      var chargingSourcesAvailable = await this.batterySensor.GetChargingSourcesAvailable();
      var response = new ChargingSourcesAvailableResponse
      {
        HomeBase = (chargingSourcesAvailable & ChargingSources.HomeBase) == ChargingSources.HomeBase,
        InternalCharger = (chargingSourcesAvailable & ChargingSources.InternalCharger) ==
                          ChargingSources.InternalCharger
      };
      return response;
    }

    public override async Task<ChargingStateResponse> GetChargingState(Empty request, ServerCallContext context)
    {
      var chargingState = await this.batterySensor.GetChargingState();
      return new ChargingStateResponse
      {
        State = (ChargingState) chargingState + 1
      };
    }
  }
}

File: RoombaPi.Client/src/app/battery/services/battery.service.ts
import { Injectable } from "@angular/core";

import { environment } from "src/environments/environment";

import { BatteryServiceClient } from "@app/grpc/battery";
import { Empty } from "google-protobuf/google/protobuf/empty_pb";

import { BatteryInfo } from "../models/batteryInfo";
import { ChargingSourcesAvailable } from "../models/chargingSourcesAvailable";
import { ChargingState } from "../models/chargingState";


@Injectable({ providedIn: "root" })
export class BatteryService {
  private readonly client: BatteryServiceClient;

  constructor() {
    this.client = new BatteryServiceClient(environment.grpcServer);
  }

  public getBatteryInfo(): Promise<BatteryInfo> {
    return new Promise<BatteryInfo>((resolve, reject) => {
      this.client.getBatteryInfo(new Empty(), null, (error, response) => {
        if (error) {
          return reject(error);
        }

        resolve(response.toObject());
      });
    });
  }

  public getChargingSourcesAvailable(): Promise<ChargingSourcesAvailable> {
    return new Promise<ChargingSourcesAvailable>((resolve, reject) => {
      this.client.getChargingSourcesAvailable(new Empty(), null, (error, response) => {
        if (error) {
          return reject(error);
        }

        resolve(response.toObject());
      });
    });
  }

  public getChargingState(): Promise<ChargingState> {
    return new Promise<ChargingState>((resolve, reject) => {
      this.client.getChargingState(new Empty(), null, (error, response) => {
        if (error) {
          return reject(error);
        }

        resolve(response.toObject().state);
      });
    });
  }
}
